File: 2501/hashset.py
from typing import *


class Solution:
    def longestSquareStreak(self, nums: List[int]) -> int:
        unique_nums = set(nums)
        ret = -1
        for num in nums:
            square = num * num
            streak = 1
            while square in unique_nums:
                streak += 1
                if ret < streak:
                    ret = streak
                square *= square
        return ret


if __name__ == '__main__':
    sol = Solution()

    nums, expected = [4,3,6,16,8,2], 3
    assert sol.longestSquareStreak(nums) == expected

    nums, expected = [2,3,5,6,7], -1
    assert sol.longestSquareStreak(nums) == expected

File: 0136/hashset.cpp
#include <gtest/gtest.h>

#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 public:
  int singleNumber(vector<int>& nums) {
    unordered_set<int> set;
    for (const auto& num : nums) {
      if (set.find(num) != set.end()) {
        set.erase(num);
      } else {
        set.insert(num);
      }
    }
    return *set.begin();
  }
};

int main() {
  Solution sol;
  vector<int> nums;
  int expected;

  nums = {2,2,1}, expected = 1;
  EXPECT_EQ(sol.singleNumber(nums), expected);

  nums = {4,1,2,1,2}, expected = 4;
  EXPECT_EQ(sol.singleNumber(nums), expected);

  nums = {2}, expected = 2;
  EXPECT_EQ(sol.singleNumber(nums), expected);
}
